include/boost/math/quadrature/gauss.hpp

```diff
@@ -74,106 +97,56 @@ template <class Real, unsigned N, unsigned Category>
 class gauss_detail;
 
 #endif
-
+#ifndef BOOST_HAS_FLOAT128
 template <class T>
 class gauss_detail<T, 7, 0>
 {
+   using storage_type = typename gauss_constant_category<T>::storage_type;
 public:
-   static std::array<T, 4> const & abscissa()
+   static std::array<storage_type, 4> const & abscissa()
    {
-      static constexpr std::array<T, 4> data = {
-         0.000000000e+00f,
-         4.058451514e-01f,
-         7.415311856e-01f,
-         9.491079123e-01f,
+      static constexpr std::array<storage_type, 4> data = {
+         static_cast<storage_type>(0.00000000000000000000000000000000000e+00L),
+         static_cast<storage_type>(4.05845151377397166906606412076961463e-01L),
+         static_cast<storage_type>(7.41531185599394439863864773280788407e-01L),
+         static_cast<storage_type>(9.49107912342758524526189684047851262e-01L),
```

Comment:
What is the reasoning behind putting have the function return a const& to a constexpr type? i.e. why can't the signature just be `constexpr`?

Reply:
I don't think functions returning arrays are allowed?  But also just because the function is marked as constexpr, does not mean that it will be evaluated as such, only that it can be.  We could probably mark the existing signatures constexpr, but I'm not sure it actually makes any practical difference - either it's inlined and optimised away or not, either which way, we can never actually call the function in a constexpr context since some overloads (for the multiprecision cases) cannot be constexpr.

Reply:
> I don't think functions returning arrays are allowed?

In c++14? That should be fine

> But also just because the function is marked as constexpr, does not mean that it will be evaluated as such, only that it can be.

If a function can be constexpr idt the compiler gets a choice on whether to make it constexpr or not. At least reading over the constexpr function spec from cppreference I don't see anything that says that constexpr function invocations are compiler defined.

> either it's inlined and optimised away or not, either which way, 

Making those constexpr removes the static here which imo does matter for user defined types

> we can never actually call the function in a constexpr context since some overloads (for the multiprecision cases) cannot be constexpr.

See the constexpr function docs section [here](https://en.cppreference.com/w/cpp/language/constexpr). For any of the overloads that can be constexpr this will be.

Reply:
>Making those constexpr removes the static here which imo does matter for user defined types

Right, we do actually want the functions static in the general case (since there is no need for a `this` pointer).  In any case, unless someone demonstrates that substantially better code can be produced with a different signature, I'll leave as is for now.